### PR TITLE
Fix:  Portable PDB MethodDebugInformation table size differs from MethodDef #3514

### PR DESCRIPTION
**Pull request by KevinRansom**
Fixes: Portable PDB MethodDebugInformation table size differs from MethodDef #3514 

When generating a portable PDB file, we failed to generate a MethodDebugInformation record for methods with no sequence points.

This PR addresses that by writing out for methods with empty sequence points a MethodDebugInformation record with
- Document = 0
- SequencePoint Blob heap index of 0

### CODE REVIEW

File: src/absil/ilwritepdb.fs [hunk 1]
```diff
@@ -330,17 +330,22 @@ let generatePortablePdb (embedAllSource:bool) (embedSourceList:string list) (sou
...
-                else 
-                    match documentIndex.TryGetValue(docs.[d].File) with
-                    | false, _ -> Unchecked.defaultof<DocumentHandle>
-                    | true, h -> h
-
-            if sps.Length = 0 then
+            let builder = new BlobBuilder()
+            builder.WriteCompressedInteger(minfo.LocalSignatureToken)
+
+            if sps = Array.empty then
```
Comment (dsyme) [inline src/absil/ilwritepdb.fs]: Minor nit - I personally feel it's better to test to check for ``sps.Length = 0`` as in the original code.  

Relying on array equality is ok, but 

(1) forces an ``equality`` constraint on the element type of the array, which may not always hold, and 
(2) may be subtle to people coming from .NET/C# where array equality is reference equality

**Comment by KevinRansom** [comment]
the fix adds a methoddebuginformation record for methods with no sequence points.  effectively it writes a 0,0 to the pdb. (unknown document, no sps)

The spec says that an index into the methoddef table can be used to index the methoddebuginformation table.

Yes I will modify the empty array check.

**Comment by KevinRansom** [comment]
It turns out that ilwrite didn't even bother to tell the pdbwriter about the sequence point free abstract methods.  (Honestly I think that was a fine choice except that the portable pdb format requires it.)

This update enables ilwrite to pass the abstract method to the pdbwrite, eliminates a few pieces of dead code.  Since the abstract method has no sequence points, this change also allow the rootscope to be an option type.

**Comment by dsyme** [comment]
@KevinRansom great work

**Comment by KevinRansom** [comment]
@dotnet-bot Windows_NT Release_ci_part2 please

**Comment by KevinRansom** [comment]
@dotnet-bot Windows_NT Release_ci_part2 please

**Comment by nosami** [comment]
I think I just encountered this bug since dogfooding portable pdb support on Mono.

I hit this assertion [here](https://github.com/mono/mono/blob/master/mono/metadata/metadata.c?utf8=%E2%9C%93#L1116). I suspect that this fixes it.

**Comment by nosami** [comment]
> I suspect that this fixes it.
This does indeed fix the debugging issue on Mono. Excellent work and perfect timing :)

Please can this be merged into fsharp/fsharp?

**Comment by nosami** [comment]
Nice work on portable pdb debugging overall too.

screen shot 2017-09-05 at 8 46 19 pm

This type of breakpoint wasn't possible before on Mono with a start column and end column (presumably because that was added to the Mono debugger and format after the F# .mdb support was implemented)

**Comment by KevinRansom** [comment]
Build keeps timing out ... I wonder why?

**Comment by KevinRansom** [comment]
I need to get to the bottom of that first.

**Comment by KevinRansom** [comment]
@nosami what is in fsharp/fsharp that is not yet in Microsoft/visualfsharp?

**Comment by nosami** [comment]
> what is in fsharp/fsharp that is not yet in Microsoft/visualfsharp?

I'm not sure, @dsyme is probably the best person to ask about that. fsharp/fsharp is the easiest repo to build on OSX/linux because of the Makefile support etc.  It's where the command line tooling for Mono is built from. I'm not even sure if it's possible to use this repo for that.

**Comment by KevinRansom** [comment]
@dotnet-bot Windows_NT Release_ci_part2 please

**Comment by dsyme** [comment]
> what is in fsharp/fsharp that is not yet in Microsoft/visualfsharp?

Practically speaking, one of the main things is a full CI testing matrix - that's really easy uing travis, I don't know how to get the nice mono installs like ``weekly``, ``nightly``, ``latest`` and so on using Jenkins.  Also the FSharp.Compiler.Tools nuget package gets built and published via CI

> Please can this be merged into fsharp/fsharp?

Please just cherry-pick it and submit a separate PR

**Comment by KevinRansom** [comment]
@dotnet-bot Windows_NT Release_ci_part2 please

**Comment by KevinRansom** [comment]
@dotnet-bot Windows_NT Release_ci_part2 please

**Comment by KevinRansom** [comment]
@dotnet-bot test this please

**Comment by dsyme** [comment]
@dotnet-bot Windows_NT Release_ci_part2 please

**Comment by KevinRansom** [comment]
@dotnet-bot Windows_NT Release_ci_part2 please

**Comment by KevinRansom** [comment]
@dotnet-bot test this please

**Comment by KevinRansom** [comment]
@dsyme, I am suspicious that : https://github.com/Microsoft/visualfsharp/commit/7791c9d83acf8fe35a1d08477e0da0f95dd758bf
is what is causing the ci to time out.

**Comment by KevinRansom** [comment]
@dotnet-bot test this please

Over the last couple of weeks we have started to see timeouts on ci_part2. If I comment out the 
````
CodeGen01,NoMT,CodeGen				CodeGen\Structure
````
in the fsharpqa suite then all of the tests past.
Or if I enable them and revert the PR: https://github.com/Microsoft/visualfsharp/commit/7791c9d83acf8fe35a1d08477e0da0f95dd758bf

They all pass, nothing about the PR, suggests this should happen, could you take a look please.

Also, the failures only seem to happen on the CI, when I run them locally they work fine.

Thx.

Kevin

**Comment by dsyme** [comment]
> Or if I enable them and revert the PR: 7791c9d

I suspect that must be coincidence, I can't see anything about that PR that would cause this.

**Comment by dsyme** [comment]
Yep, still timing out

**Comment by KevinRansom** [comment]
@dotnet-bot test Windows_NT Release_ci_part2 Build please

**Comment by KevinRansom** [comment]
@dotnet-bot test Windows_NT Release_ci_part2 Build please

**Comment by KevinRansom** [comment]
@dotnet-bot test Windows_NT Release_ci_part2 Build please

**Comment by KevinRansom** [comment]
@dotnet-bot test Windows_NT Release_ci_part2 Build please

**Comment by KevinRansom** [comment]
@dotnet-bot test this please

**Comment by KevinRansom** [comment]
@dotnet-bot test Windows_NT Release_ci_part2 Build  please

**Pull request merged**